File: Stacks/10/author/validParen.js
var isValid = function (s) {
  if (s.length === 0) return true;

  const parens = {
    '{': '}',
    '(': ')',
    '[': ']',
  };

  const stack = [];

  for (let ch of s) {
    if (parens[ch]) {
      stack.push(ch);
    } else {
      leftParen = stack.pop();
      if (parens[leftParen] !== ch) {
        return false;
      }
    }
  }
  return stack.length === 0;
};

const ch = ']';
console.log(isValid(ch));

File: Stacks/10/author/validParen.py

def isValid(s):
    if len(s) == 0: return True

    parens = {
        '(':')',
        '{':'}',
        '[':']'
    }
    stack = []

    for ch in s:
        if parens.get(ch):
            stack.append(ch)
        else:
            if len(stack):
                leftParen = stack.pop()
                if parens[leftParen] != ch:
                    return False
            else:
                return False
    return len(stack) == 0


ch = '([])'
print(isValid(ch))
